_episodes/l3_03_test_fixture.md

```diff
@@ -0,0 +1,505 @@
+---
+## 1. Brief Introduction of Test Fixtures
+
+A test fixture, in the context of unit testing, refers to the preparation and configuration needed to run a set of test cases. It includes the setup of the test environment, the creation of necessary objects or resources, and the cleanup steps after the tests are executed. Test fixtures help ensure that the tests are performed in a controlled and consistent environment, providing reliable and reproducible results.
+
+In GoogleTest, a test fixture is implemented using a test fixture class. This class serves as a container for shared setup and cleanup logic, as well as any shared objects or resources required by the test cases within the fixture.
+
+### 1.1 When is a Test Fixture Needed?
+
+A test fixture is typically used in the following scenarios:
+
+1. **Shared Setup and Teardown**: When multiple test cases require a common set of setup or teardown (which means cleaning in general and is the opposite of setup) steps, a test fixture is beneficial. Instead of duplicating the setup and teardown code in each test case, we can define it once in the test fixture and reuse it across all the tests.
+
+2. **Reducing Code Duplication**: Test fixtures help in reducing code duplication. By encapsulating the common setup and teardown logic within a fixture, we avoid duplicating the same code in multiple test cases. This improves code maintainability and reduces the chances of errors due to inconsistent or incomplete setup/teardown.
+
+3. **Isolation and Independence**: Test fixtures provide a level of isolation and independence for each test case. Each test case within a fixture runs in its own instance of the fixture class, ensuring that changes made by one test case do not affect the others. This allows for parallel execution of test cases without interference.
+
+Let us understand how to create a test fixture, setup and teardown functions with examples.
+
+## 2. Problem under consideration for testing
+
+In order to create our own test fixture, we will first explain the context or problem that we are trying to solve. We will be adding tests to check that our code works as intended by making use of test fixtures.
+
+Consider that you want to write a program to manage the details of an employee. The program should allow you to add basic details of employee such as:
+
+1. Name
+2. Age
+3. Basic Salary
+4. Number of years of employment
+5. Basic Bonus the employee has received in this year.
+
+The program should calculate the Net Bonus, Tax and Salary based on the following rules.
+
+1. Bonus Rule: An employee gets an additional 1000£ bonus if she or he has worked for more than 10 years.
+2. Tax Rule: Tax is calculated on a combination of basic salary and net bonus as shown below.
+
+    * `0` if salary is less than `10k` GBP.
+    * `10%` for salary between `10K-20K` GBP.
+    * `20%` for salary between `20K-50K` GBP
+    * `50%` for salary greater than `50K` GBP
```

Comment:
Maybe we should use the actual values?

```suggestion
    * `0` if salary is less than £12,570
    * `20%` for salary between `£12,571 and £50,270
    * `40%` for salary between £50,271 and £125,140
    * `45%` for salary greater than £125,140
```

Reply:
No, I don't even worry about that. The actual values have no relation to the course. Moreover, it will force me to do some more calculations by hand for the tests. I will keep this as they are.